pkg/cloud/services/ec2/instances.go

```diff
@@ -1138,6 +1149,21 @@ func getCapacityReservationSpecification(capacityReservationID *string) *ec2.Cap
 	}
 }
 
+func getCapacityBlocksInstanceMarketOptionsRequest(capacityBlock *bool, capacityReservationID *string) (*ec2.InstanceMarketOptionsRequest, error) {
+	if !*capacityBlock {
+		// Instance is not capacity-blocks
+		return nil, nil
+	}
+	if capacityReservationID == nil {
+		return nil, errors.Errorf("Missing capacityReservationID, CapacityBlock is set but capacityReservationID is empty")
+	}
+
+	instanceMarketOptionsRequest := &ec2.InstanceMarketOptionsRequest{}
+	instanceMarketOptionsRequest.SetMarketType(ec2.MarketTypeCapacityBlock)
+
+	return instanceMarketOptionsRequest, nil
+}
+
```

Comment:
How about including this in the `getInstanceMarketOptionsRequest` below like so:
Beware: Untested code

```go
func getInstanceMarketOptionsRequest(i *infrav1.Instance) (*ec2.InstanceMarketOptionsRequest, error) {
   // Handle Capacity Block case.
   if ptr.Deref(i.CapacityBlock, false) {
       if i.CapacityReservationID == nil {
           return nil, fmt.Errorf("capacityReservationID is required when CapacityBlock is enabled")
       }
       return &ec2.InstanceMarketOptionsRequest{
           MarketType: aws.String(ec2.MarketTypeCapacityBlock),
       }, nil
   }

   // Handle Spot instance case.
   if i.SpotMarketOptions == nil {
       // Instance is not a Spot instance
       return nil, nil
   }

   // Set required values for Spot instances
   spotOpts := &ec2.SpotMarketOptions{
       // The following two options ensure that:
       // - If an instance is interrupted, it is terminated rather than hibernating or stopping
       // - No replacement instance will be created if the instance is interrupted
       // - If the spot request cannot immediately be fulfilled, it will not be created
       // This behaviour should satisfy the 1:1 mapping of Machines to Instances as
       // assumed by the Cluster API.
       InstanceInterruptionBehavior: aws.String(ec2.InstanceInterruptionBehaviorTerminate),
       SpotInstanceType:            aws.String(ec2.SpotInstanceTypeOneTime),
   }
   
   maxPrice := ptr.Deref(i.SpotMarketOptions.MaxPrice, "")
   if maxPrice != "" {
       spotOpts.MaxPrice = aws.String(maxPrice)
   }

   return &ec2.InstanceMarketOptionsRequest{
       MarketType:  aws.String(ec2.MarketTypeSpot),
       SpotOptions: spotOpts,
   }, nil
}
```

And then invoke it this way:
```go
marketOptions, err := getInstanceMarketOptions(i)
if err != nil {
    return nil, err
}
if marketOptions != nil {
    input.InstanceMarketOptions = marketOptions
}
```

Reply:
yes, I agree with it. I thought of creating separate method then it justify the single responsibility principle.
But, this approach makes sense too. Changes are updated